### PR TITLE
[Newsletter] Support newsletter resubscription after unsubscribing using DOI

### PR DESCRIPTION
**Pull request by apfelbox**
### 1. Why is this change necessary?
Right now, when using DOI for newsletter subscriptions, the user can't resubscribe.

### 2. What does this change do, exactly?
When DOI is confirmed, internally the `confirmedAt` field is set. 
Now when subscribing a user, if the user already exists and has a `confirmedAt` set, then the request is aborted.

That is however incorrect in the current implementation: because we never reset the `confirmedAt` when unsubscribing the user, an unsubscribed user will also early exit here and can't resubscribe.

### 3. Describe each step to reproduce the issue or behaviour.
See the test case:

1. subscribe
2. confirm DOI
3. unsubscribe
4. try to resubscribe

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

---

A different solution would be to reset the `confirmedAt` field when unsubscribing, however this field might be relevant if looking up old users ("when did I approve the confirmation? Oh, at 5th of May, ok")

A second solution might be to add a task that continuously removes opt-out newsletter subscriptions, but that might be unwanted for certain installations.

### CODE REVIEW

**Comment by apfelbox** [comment]
The test is pretty slow right now, as we need to hit 4 API endpoints consecutively.. Any tips on how to speed that up are highly appreciated.

**Comment by J-Rahe** [comment]
Hey @apfelbox, you dont actually have to make 4 API calls, instead you could jut put an user who has been registered with double opt-in into the database at the beginnig of your test.

Also i'm currently waiting for an answer regarding the confirmedAt Field.
I will come back to you once i have an answer.

**Comment by apfelbox** [comment]
Ah, I wasn't sure on whether it is fine to directly insert these rows via query.
I will wait for your feedback regarding just resetting `confirmedAt` and then adapt the PR accordingly. 👍

File: changelog/_unreleased/2021-06-15-allow-newsletter-resubscribe.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+title: Allow newsletter resubscribe after opt-out
+author: Jannik Zschiesche
+author_email: j.zschiesche@21torr.com
+author_github: @apfelbox
+---
+# Core
+* Resubscribing to the newsletter after opt-out (using DOI) is now properly supported.
```
Comment (J-Rahe) [inline changelog/_unreleased/2021-06-15-allow-newsletter-resubscribe.md]: please make use of the [keywords](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md)

File: src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php [hunk 2]
```diff
@@ -108,6 +108,71 @@ public function testSubscribe(): void
...
+        $row = $connection->fetchAssoc('SELECT * FROM newsletter_recipient WHERE email = "test@test.de"');
+        static::assertSame('optIn', $row['status']);
+
+        // 3: unsubscribe
+        $this->browser
+            ->request(
+                'POST',
+                '/store-api/newsletter/unsubscribe',
+                [
+                    'email' => 'test@test.de',
```
Comment (J-Rahe) [inline src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php]: please also change the mail to an example.com (just to make sure we don't accidentally send mails to the owner).

File: src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php [hunk 3]
```diff
@@ -108,6 +108,71 @@ public function testSubscribe(): void
...
+                    'option' => 'subscribe',
+                    'storefrontUrl' => 'http://localhost',
+                ]
+            );
+
+        static::assertTrue($this->browser->getResponse()->isSuccessful());
+        $row = $connection->fetchAssoc('SELECT * FROM newsletter_recipient WHERE email = "test@test.de"');
+        static::assertSame('notSet', $row['status']);
+    }
+
```
Comment (J-Rahe) [inline src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php:182]: also please confirm the subscription an second time.

**Review by J-Rahe — CHANGES_REQUESTED** [review]
Hey @apfelbox, you can overwrite the confirmedAt field on the second confirmation. I have also added some comments regarding some small changes.

File: src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php (hunk 3 above)
Comment (apfelbox) [reply src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php:182]: That is a great idea, as this fails currently. The issue is, that `NewsletterConfirmRouter::confirm()` expects `confirmedAt` to be `null`.
So there probably is no other way than to reset the `confirmedAt` to null when unsubscribing.

Comment (J-Rahe) [reply src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php:182]: why not remove the check from the 'getBeforeConfirmSubscribeValidation'? It is private and prevents an new confirmation. (It would be better to keep the confirmedAt until someone subscribes again)

File: src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php (hunk 2 above)
Comment (apfelbox) [reply src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php]: I changed all email addresses in the test file

File: changelog/_unreleased/2021-06-15-allow-newsletter-resubscribe.md (hunk 1 above)
Comment (apfelbox) [reply changelog/_unreleased/2021-06-15-allow-newsletter-resubscribe.md]: I hope it's correct now

**Comment by apfelbox** [comment]
I tried inserting the entry via an SQL statement, but I would need to fetch a lot of different IDs ( `language_id`, `sales_channel_id`) due to foreign keys, so it would be quite a lot of code. Is there some helper for that?

File: src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php (hunk 3 above)
Comment (apfelbox) [reply src/Core/Content/Test/Newsletter/SalesChannel/NewsletterSubscribeRouteTest.php:182]: All right. I removed it from the validation constraint collection and restored the previous version.
I also extended the test to check that the confirmation date is null / not null and that it changes after confirming a second time.

**Comment by J-Rahe** [comment]
There is the IntegrationTestBehaviour which includes the BasicTestDataBehaviour so you can just use e.g. this->getValidSalutationId.

**Comment by apfelbox** [comment]
@J-Rahe thanks for the tip, I think the implementation is quite nice now.
Everything should be fixed 👍

**Comment by mitelg** [comment]
thanks for your contribution @apfelbox   👍🎉💙
